### PR TITLE
DOCSP-6442: Parse repo-wide substitutions

### PR DESCRIPTION
**Pull request by sophstad**
[[JIRA](https://jira.mongodb.org/browse/DOCSP-6442)] Parse substitutions defined in the `snooty.toml` under `[substitutions]` and include their nodes in the AST.

### CODE REVIEW

File: snooty/parser.py [hunk 1]
```diff
@@ -536,6 +552,16 @@ def __init__(
...
         }
 
+        substitution_nodes: Dict[str, SerializableType] = {}
+        for k, v in self.config.substitutions.items():
+            node, diagnostics = parse_rst(self.parser, root, v)
+            substitution_nodes[k] = node.ast
+
+        self.config.substitution_nodes = substitution_nodes
+
+        # Re-parse using our updated config that now includes substitutions represented as nodes
```
Comment (i80and) [inline snooty/parser.py]: Is this actually required?

File: snooty/types.py [hunk 2]
```diff
@@ -260,6 +260,8 @@ class ProjectConfig:
     name: str
     source: str = field(default="source")
     constants: Dict[str, object] = field(default_factory=dict)
+    substitutions: Dict[str, str] = field(default_factory=dict)
+    substitution_nodes: Dict[str, SerializableType] = field(default_factory=dict)
```
Comment (i80and) [inline snooty/types.py]: I'd have a comment on this attribute noting that it's created after parse time

File: snooty/test_parser.py [hunk 3]
```diff
@@ -244,6 +244,65 @@ def test_admonition() -> None:
...
+|yaml| ain't markup language.
+""",
+    )
+    assert diagnostics == []
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <paragraph>
+        <substitution_reference>
+        <paragraph>
```
Comment (i80and) [inline snooty/test_parser.py]: I don't think the substituted text should be in its own paragraph; we should peel that off as well

File: snooty/types.py [hunk 4]
```diff
@@ -260,6 +260,9 @@ class ProjectConfig:
     name: str
     source: str = field(default="source")
     constants: Dict[str, object] = field(default_factory=dict)
+    substitutions: Dict[str, str] = field(default_factory=dict)
+    # This variable is populated after parse time with the nodes from `substitutions` strings
```
Comment (i80and) [inline snooty/types.py]: Well, more accurately, it's more along the lines of
```
substitution_nodes contains a parsed representation of the substitutions member, and is populated on Project initialization.
```

File: snooty/test_parser.py [hunk 5]
```diff
@@ -244,6 +244,63 @@ def test_admonition() -> None:
...
+        """
+|yaml| ain't markup language.
+""",
+    )
+    assert diagnostics == []
+    check_ast_testing_string(
+        page.ast,
+        """<root>
+        <paragraph>
+        <substitution_reference>
```
Comment (i80and) [inline snooty/test_parser.py]: The `<substitution_reference>` should be gone too, shouldn't it?

Comment (sophstad) [reply snooty/test_parser.py]: At the moment, we're taking the child nodes returned by `parse_rst()` and assigning them to the `children` property of the substitution reference node. With this approach, I believe the `<substitution_reference>` tag should still be present. I also think that keeping it in the AST can help with debugging on the front-end. But if you feel that it isn't logical to include it in the AST, I'm open to that approach as well!

**Pull request merged**